src/coprocessor/codec/mysql/duration.rs

```diff
@@ -32,13 +42,165 @@ fn check_dur(dur: &StdDuration) -> Result<()> {
     Ok(())
 }
 
-fn tm_to_secs(t: Tm) -> u64 {
-    t.tm_hour as u64 * SECS_PER_HOUR + t.tm_min as u64 * SECS_PER_MINUTE + t.tm_sec as u64
+#[inline]
+fn check_hour(hour: u32) -> Result<u32> {
+    if hour > MAX_HOURS {
+        Err(invalid_type!(
+            "invalid hour value: {} larger than {}",
+            hour,
+            MAX_HOURS
+        ))
+    } else {
+        Ok(hour)
+    }
+}
+
+#[inline]
+fn check_minute(minute: u32) -> Result<u32> {
+    if minute > MAX_MINUTES {
+        Err(invalid_type!(
+            "invalid minute value: {} larger than {}",
+            minute,
+            MAX_MINUTES
+        ))
+    } else {
+        Ok(minute)
+    }
+}
+
+#[inline]
+fn check_second(second: u32) -> Result<u32> {
+    if second > MAX_SECONDS {
+        Err(invalid_type!(
+            "invalid second value: {} larger than {}",
+            second,
+            MAX_SECONDS
+        ))
+    } else {
+        Ok(second)
+    }
+}
+
+fn buf_to_int(buf: &[u8]) -> u32 {
+    buf.iter().fold(0, |acc, c| acc * 10 + u32::from(c - b'0'))
 }
 
+// Functionality:
+// Extract a `u32` from a buffer which matches pattern: `\d+.*`
+//
+// The range of MySQL's TIME is `-838:59:59 ~ 838:59:59`, so we can at most read 7 digits
+// (pattern like `HHHMMSS`)
+named!(
+    read_int<u32>,
+    map_res!(digit1, |buf: &[u8]| if buf.len() > 7 {
+        Err(invalid_type!("invalid time value, more than {} digits", 7))
+    } else {
+        Ok(buf_to_int(buf))
+    })
+);
+
+// Functionality:
+// Extract a `u32` with `fsp` from a buffer which matches pattern: `\d+.*`
```

Comment:
"with fsp" is still not clear. Please add a simple example, what we will get from this function if the string to parse is 123456 and fsp is 4.